### PR TITLE
Avoid duplicate clauses when using #or

### PR DESCRIPTION
**Pull request by MaxLap**
Attempt number 2 for this change. A refactor of WhereClause by @sgrif made this change a lot simpler. Initial proposal at #29805.

Right now, when using `#or`, nothing is done to try to avoid duplicated conditions. As a result, query length increases at an exponential rate. This makes it really hard to parse for humans, making debugging / understanding the query log really hard.

To show the scale of the problem:
```
class Project
    scope :big, -> { where(big1: true).or(where(big2: true)) }
    scope :important, -> { where(important1: true).or(where(important2: true)) }
    scope :ongoing, -> { where(ongoing1: true).or(where(ongoing2: true)) }
end

Project.big # 2 conditions
#=> SELECT...WHERE ("projects"."big1" = ? OR "projects"."big2" = ?) 
Project.big.important # 6 conditions
#=> SELECT...WHERE (("projects"."big1" = ? OR "projects"."big2" = ?) AND "projects"."important1" = ? OR ("projects"."big1" = ? OR "projects"."big2" = ?) AND "projects"."important2" = ?)
Project.big.important.ongoing # 14 conditions
#=> To long to show
# Then 30 conditions
# Then 62
# Then 126 conditions, for using #or 6 times. This should be 12!

# After this fix, it is correctly 2, 4, 6, 8, 10, ... conditions.
```

### CODE REVIEW

File: activerecord/test/cases/relation/where_clause_test.rb [hunk 1]
```diff
@@ -181,6 +181,39 @@ class WhereClauseTest < ActiveRecord::TestCase
...
+        WhereClause.new([table["col_#{i}"].eq(bind_param(i))])
+      end
+
+      actual = wcs[0]
+      expected = wcs[0]
+
+      actual = (actual + wcs[1]).or(actual + wcs[2])
+      expected = expected + wcs[1].or(wcs[2])
+
+      actual = (actual + wcs[3] + wcs[4]).or(actual + wcs[5] + wcs[6])
```
Comment (lugray) [inline activerecord/test/cases/relation/where_clause_test.rb]: Is this addtional manipulation really necessary for the test?

File: activerecord/test/cases/relation/where_clause_test.rb [hunk 2]
```diff
@@ -181,6 +181,39 @@ class WhereClauseTest < ActiveRecord::TestCase
...
+    test "or places common conditions before the OR" do
+      wcs = (0..6).map do |i|
+        WhereClause.new([table["col_#{i}"].eq(bind_param(i))])
+      end
+
+      actual = wcs[0]
+      expected = wcs[0]
+
+      actual = (actual + wcs[1]).or(actual + wcs[2])
+      expected = expected + wcs[1].or(wcs[2])
```
Comment (lugray) [inline activerecord/test/cases/relation/where_clause_test.rb]: The redefinition makes this harder to understand.  Why not instead:

```ruby
common = wcs[0]

actual = (common + wcs[1]).or(common + wcs[2])
expected = common + wcs[1].or(wcs[2])
```

File: activerecord/lib/active_record/relation/where_clause.rb [hunk 3]
```diff
@@ -26,15 +32,16 @@ def except(*columns)
...
-        else
-          WhereClause.new(
-            [ast.or(other.ast)],
-          )
-        end
+        left = self - other
+        common = self - left
+        right = other - common
+
+        return common if left.empty? || right.empty?
```
Comment (sgrif) [inline activerecord/lib/active_record/relation/where_clause.rb]: I'd prefer to avoid postfix conditionals (and early returns for that matter) if possible. How about this instead?

```ruby
if left.empty? || right.empty?
  common
else
  or_clause = WhereClause.new(
    [left.ast.or(right.ast)],
  )
  common + or_clause
end
```

File: activerecord/test/cases/relation/where_clause_test.rb [hunk 4]
```diff
@@ -181,6 +181,39 @@ class WhereClauseTest < ActiveRecord::TestCase
       assert_equal WhereClause.empty, WhereClause.empty.or(where_clause)
     end
 
+    test "or places common conditions before the OR" do
+      wcs = (0..6).map do |i|
+        WhereClause.new([table["col_#{i}"].eq(bind_param(i))])
```
Comment (sgrif) [inline activerecord/test/cases/relation/where_clause_test.rb]: This sort of thing makes this test really difficult to read. How about something like this?

```ruby
a = WhereClause.new(
  [table["id"].eq(bind_param(1)), table["name"].eq(bind_param("Sean"))],
)
b = WhereClause.new(
  [table["id"].eq(bind_param(1)) ,table["hair_color"].eq(bind_param("black"))],
)
common = WhereClause.new(
  [table["id"].eq(bind_param(1))],
)
or_clause = WhereClause.new(
  [table["name"].eq(bind_param("Sean")).or(table["hair_color"].eq(bind_param("black")))],
)

assert_equal common + or_clause, a.or(b)
```

File: activerecord/test/cases/relation/where_clause_test.rb [hunk 5]
```diff
@@ -181,6 +181,39 @@ class WhereClauseTest < ActiveRecord::TestCase
...
+
+    test "or will use common conditions only if one side only has common conditions" do
+      common = WhereClause.new(
+        [table["id"].eq(bind_param(1)), "foo = bar"]
+      )
+
+      extra = WhereClause.new([table["extra"].eq(bind_param("pluto"))])
+
+      actual_extra_only_on_left = (common + extra).or(common)
+      actual_extra_only_on_right = (common).or(common + extra)
```
Comment (sgrif) [inline activerecord/test/cases/relation/where_clause_test.rb]: I also find this really hard to read. Can we either just test this once, or just write out the actual conditions that we're testing?

**Review by sgrif — CHANGES_REQUESTED** [review]

**Comment by MaxLap** [comment]
Personally, I thought that using variables and then doing + to concatenate conditions together helpers with readability. But i'll change it to be more basic and will split the tests into smaller ones.

File: activerecord/test/cases/relation/where_clause_test.rb (hunk 4 above)
Comment (MaxLap) [inline activerecord/test/cases/relation/where_clause_test.rb]: Just tried, that doesn't work because WhereClause#or isn't call for both sides, and WhereClause#or always adds a And node around the predicates, even if there only one.

I'll tweak the test.

**Comment by MaxLap** [comment]
I did the changes and added another test.
